src/main/java/org/opensearch/plugin/insights/core/service/QueryGroupingService.java

```diff
@@ -0,0 +1,243 @@
+     */
+    private PriorityQueue<SearchQueryRecord> maxHeapQueryStore;
+
+    /**
+     * Top N size based on the configuration set
+     */
+    private int topNSize;
+
+    public QueryGroupingService(
+        MetricType metricType,
+        GroupingType groupingType,
+        AggregationType aggregationType,
+        PriorityQueue<SearchQueryRecord> topQueriesStore,
+        int topNSize
+    ) {
+        this.groupingType = groupingType;
+        this.metricType = metricType;
+        this.aggregationType = aggregationType;
+        this.groupIdToAggSearchQueryRecord = new HashMap<>();
+        this.minHeapTopQueriesStore = topQueriesStore;
+        this.maxHeapQueryStore = new PriorityQueue<>((a, b) -> SearchQueryRecord.compare(b, a, metricType));
+
+        this.topNSize = topNSize;
+    }
+
+    /**
+     * Add query to the group based on the GroupType setting.
+     * The grouping of metrics will be stored within the searchQueryRecord.
+     * @param searchQueryRecord record
+     * @return return the search query record that represents the group
+     */
+    public SearchQueryRecord addQueryToGroup(SearchQueryRecord searchQueryRecord) {
+        if (groupingType == GroupingType.NONE) {
+            throw new IllegalArgumentException("Do not use addQueryToGroup when GroupingType is None");
+        }
+        SearchQueryRecord aggregateSearchQueryRecord;
+        String groupId = getGroupingId(searchQueryRecord);
+
+        // New group added to the grouping service
+        // Add to min PQ and overflow records to max PQ (if the number of records in the min PQ exceeds the configured size N)
+        if (!groupIdToAggSearchQueryRecord.containsKey(groupId)) {
+            aggregateSearchQueryRecord = searchQueryRecord;
+            aggregateSearchQueryRecord.setGroupingId(groupId);
+            aggregateSearchQueryRecord.setMeasurementAggregation(metricType, aggregationType);
+            addToMinPQOverflowToMaxPQ(aggregateSearchQueryRecord, groupId);
+        }
+        // Existing group being updated to the grouping service
+        // 1. If present in min PQ
+        // - remove the record from the min PQ
+        // - update the aggregate record (aggregate measurement could increase or decrease)
+        // - If max PQ contains elements, add to max PQ and promote any records to min PQ
+        // - If max PQ is empty, add to min PQ and overflow any records to max PQ
+        // 2. If present in max PQ
+        // - remove the record from the max PQ
+        // - update the aggregate record (aggregate measurement could increase or decrease)
+        // - If min PQ is full, add to min PQ and overflow any records to max PQ
+        // - else, add to max PQ and promote any records to min PQ
+        else {
+            aggregateSearchQueryRecord = groupIdToAggSearchQueryRecord.get(groupId).v1();
+            boolean isPresentInMinPQ = groupIdToAggSearchQueryRecord.get(groupId).v2();
+            if (isPresentInMinPQ) {
```

Comment:
How are we doing the aggregation on measurement? Is it just the sum of all measurements or something else? Also, how are we adhering to the running window constraint if we aggregate the measurements together?

Reply:
Its based on the AggregationType (NONE, AVG, SUM)